test/testdata/bound_proc.rb

```diff
@@ -0,0 +1,63 @@
+# typed: true
+
+module Concern
+  def included(&block)
+    @block = block
+  end
+end
+
+module Readable
+  extend Concern
+
+  included do
+    before_create :do_this
+  # ^^^^^^^^^^^^^^^^^^^^^^ error: Method `before_create` does not exist on `T.class_of(Readable)`
+  end
+end
+
+module Writable
+  extend Concern
+
+  included do
+    T.bind(self, T.any(T.class_of(Article), T.class_of(Post)))
+    before_create :do_this
+  # ^^^^^^^^^^^^^^^^^^^^^^ error: Method `before_create` does not exist on `T.class_of(Article)` component of `T.any(T.class_of(Article), T.class_of(Post))`
+  end
+end
+
+module Shareable
+  extend Concern
+
+  included do
+    T.bind(self, T.class_of(Article))
+    before_save :do_this
+  end
+end
+
+class Post
+  def self.before_create(name); end
+
+  include Readable
+  include Writable
+
+  PROC = lambda do
+    author
+  # ^^^^^^ error: Method `author` does not exist on `T.class_of(Post)`
+  end
+
+  def author; end
+end
+
+class Article
+  def self.before_save(something); end
+
+  include Writable
+  include Shareable
+
+  PROC = lambda do
+    T.bind(self, Article)
```

Comment:
Can we also add a test showing the expected behaviour when passing something else than `self` to `T.bind` please 🙏

Reply:
Added that case.